src/fabric/executor/multi_threaded/result_buffer.rs

```diff
@@ -1,152 +1,282 @@
-//! Contains a concurrent-safe result buffer based on `DashMap` and
-//! `AtomicCell`s
+//! Defines a buffer for storing the results of a multi-threaded executor
 //!
-//! Laying the `AtomicCell` into the `DashMap` allows us to fully realize the
-//! concurrency of the write-once read-many pattern, as we can release the read
-//! lock on the `DashMap` immediately and read the value thereafter
+//! The design of the buffer is specifically tailored to the application via the
+//! following assumptions it is aware of:
+//! - A given entry in the buffer is only ever written once
+//! - Values are only read after they are written, meaning a separate mechanism
+//!   accounts values as being written
+//! - The size of the buffer may not be known when it is allocated
+//!
+//! As a result we use a linked list of shard vectors to store the results. Each
+//! vector holds `AtomicCell`s that may be dereferenced directly to get the
+//! result. The linked list allows us to grow the buffer as needed without
+//! needing to copy the existing results or reallocate the entire buffer
+//!
+//! The size hint passed into the constructor should be used to tailor the size
+//! of the buffer to the expected number of results. This is used to
+//! pre-allocate the first shard vector. Performance sensitive applications
```

Comment:
The size hint is also used to fix the size of all new shards that are added to the buffer, that should be mentioned here